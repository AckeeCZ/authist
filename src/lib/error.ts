export enum ErrorCode {
    PasswordMismatch = 'You entered a wrong password!',
    EmailRequired = 'Email is required field, please enter the email!',
    PasswordRequired = 'Password is required field, please enter the password!',
    UserNotFound = "It seems this user doesn't exist, are you sure you entered correct credentials?",
    AuthenticationRequired = 'You are not authenticated!',
    UnsupportedAuthorization = 'Unsupported authorization type!',
    MissingRefreshToken = 'Refresh token is required parameter! Please send it in body or query as refreshToken.',
    UpdatePasswordNotImplemented = "Function 'updatePassword' is not implemented!" +
        "Please add 'updatePassword' functionality under the 'emailPassword' provider options.",
    ResetPasswordTokenRequired = 'Reset password token is required parameter! Please send it in body or query as token.',
    InvalidResetPasswordToken = 'Reset password token is not valid, some mandatory data are missing!',
}

class GenericError extends Error {
    constructor(
        public readonly message: string,
        public readonly status: number,
        public readonly errorCode: string = ''
    ) {
        super(message);
        Object.setPrototypeOf(this, new.target.prototype);
        this.name = this.constructor.name;
    }

    public toJSON() {
        return {
            message: this.message,
            status: this.status,
            errorCode: this.errorCode,
            errorClass: this.constructor.name,
            stack: this.stack,
        };
    }
    static apiErrorData = (error?: typeof ErrorCode[keyof typeof ErrorCode]) => {
        const { message, code } = (error as any) || { message: '', code: '' };
        return [code, message] as [string, string];
    }
}

export class NotAuthenticated extends GenericError {
    constructor(error?: typeof ErrorCode[keyof typeof ErrorCode]) {
        const [code, message] = GenericError.apiErrorData(error);
        super(message, 401, code);
    }
}

const mapErrorCodeValues = (): typeof ErrorCode => {
    const result: any = {};
    const eCode = Object(ErrorCode);
    for (const key in ErrorCode) {
        result[key] = {
            code: key,
            message: eCode[key],
        };
    }
    return result;
};

export const ERROR_CODE = mapErrorCodeValues();
