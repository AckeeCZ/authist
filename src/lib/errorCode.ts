enum ErrorCode {
    PasswordMismatch = 'PASSWORD_MISMATCH',
    UsernameRequired = 'USERNAME_REQUIRED',
    PasswordRequired = 'PASSWORD_REQUIRED',
    UserNotFound = 'USER_NOT_FOUND',
    InvalidToken = 'INVALID_TOKEN',
}

export default ErrorCode;
