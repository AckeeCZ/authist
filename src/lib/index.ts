export {
    Authenticator,
    AuthistOptions,
    Credentials,
    createAuthenticator,
    TokenOptions,
    User,
    UserCredentials,
    UserInfo,
    UserMetadata,
} from './authist';
export { ERROR_CODE, ErrorCode, NotAuthenticated } from './error';
export { BearerError, bearer, expressBearer, expressChangePassword, expressResetPassword } from './authMiddleware';
export { verifyToken, createCredentials } from './credentialsService';
export * as emailPasswordProvider from './providers/emailPasswordProvider';
export * as facebookProvider from './providers/facebookProvider';
export * as googleProvider from './providers/googleProvider';
export * as resetPasswordService from './resetPasswordService';
