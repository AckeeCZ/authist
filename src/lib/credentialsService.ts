import { sign, verify } from 'jsonwebtoken';
import { AuthistOptions, Credentials, User } from './authist';
import { ERROR_CODE, NotAuthenticated } from './error';
import { handleError } from './providerUtils';

export const createCredentials = (user: User, options: AuthistOptions): Credentials => {
    const getPayload = (tokenType: string) => ({
        user: {
            uid: user.uid,
        },
        type: tokenType,
    });
    return {
        accessToken: sign(getPayload('accessToken'), getSecret(options), {
            expiresIn: (options.token?.lifetimeMinutes || 60) * 60,
        }),
        refreshToken: sign(getPayload('refreshToken'), getSecret(options), {
            expiresIn: (options.token?.refreshLifetimeMinutes || 60 * 24 * 60) * 60,
        }),
        expiresIn: (options.token?.lifetimeMinutes || 60) * 60,
        refreshExpiresIn: (options.token?.refreshLifetimeMinutes || 60 * 24 * 60) * 60,
    };
};

export const verifyToken = async (token: string, options: AuthistOptions) => {
    try {
        if (!options.getUserById) {
            throw new Error('Function `getUserById` is not implemented!');
        }
        const secret = getSecret(options, true);
        const decodedToken = verify(token, secret) as { user: { uid: string } };
        const { uid } = decodedToken.user;
        const user = await options.getUserById(uid);
        if (!user) {
            throw new NotAuthenticated(ERROR_CODE.UserNotFound);
        }
        return user;
    } catch (error) {
        return handleError(error, {}, options);
    }
};

export const getSecret = (options: AuthistOptions, verify = false) => {
    if (verify && options.token?.jwtPublicKey) {
        return options.token.jwtPublicKey;
    }
    if (options.token?.jwtPrivateKey) {
        return options.token.jwtPrivateKey;
    }
    if (options.token?.jwtSecret) {
        return options.token.jwtSecret;
    }
    console.warn('You should set the JWT secret or private / public keys to prevent security breaches.');
    return 'secret';
};

export const refreshToken = (options: AuthistOptions) => async (refreshToken: string) => {
    const user = await verifyToken(refreshToken, options);
    const credentials = await createCredentials(user, options);
    return { credentials, user };
};
