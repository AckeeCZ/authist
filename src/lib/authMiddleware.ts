import { NextFunction, Request, Response } from 'express';
import { AuthistOptions, User } from './authist';
import { verifyToken } from './credentialsService';
import { ERROR_CODE, NotAuthenticated } from './error';

export type BearerError = Error & { status?: number; statusCode?: number; toJSON?: any };

export const bearer = async (
    options: AuthistOptions,
    authorization?: string
): Promise<[BearerError | undefined, User | undefined]> => {
    if (!authorization) {
        return [new NotAuthenticated(ERROR_CODE.AuthenticationRequired), undefined];
    }
    const [bearer, accessToken] = authorization.split(' ');
    if (!bearer || bearer.toLowerCase() !== 'bearer') {
        return [new NotAuthenticated(ERROR_CODE.UnsupportedAuthorization), undefined];
    }
    try {
        const user = await verifyToken(accessToken, options);
        return [undefined, user];
    } catch (error) {
        return [error, undefined];
    }
};

export const expressBearer = (options: AuthistOptions) => async (
    request: Request & { user?: User },
    response: Response,
    next: NextFunction
) => {
    const [error, user] = await bearer(options, request.headers.authorization);
    if (error) {
        if (options.onExpressAuthenticationFailure) {
            return options.onExpressAuthenticationFailure(error, request, response, next);
        }
        response.statusCode = error.status ?? error.statusCode ?? 401;
        return response.json({ message: error?.message });
    }
    request.user = user;
    next();
};
