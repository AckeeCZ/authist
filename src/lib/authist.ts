import { NextFunction, Request, Response } from 'express';
import { expressBearer } from './authMiddleware';
import { EmailPasswordProviderOptions, signInWithEmailAndPassword } from './providers/emailPasswordProvider';
import { FacebookProviderOptions, signInWithFacebook } from './providers/facebookProvider';
import { GoogleSignInOptions, signInWithGoogle } from './providers/signInWithGoogle';

export const createAuthenticator = (options: AuthistOptions): Authenticator => ({
    expressBearer: expressBearer(options),
    signInWithEmailAndPassword: signInWithEmailAndPassword(options),
    signInWithFacebook: signInWithFacebook(options),
    signInWithGoogle: signInWithGoogle(options),
});

export interface Authenticator {
    expressBearer: ReturnType<typeof expressBearer>;
    signInWithEmailAndPassword: ReturnType<typeof signInWithEmailAndPassword>;
    signInWithFacebook: ReturnType<typeof signInWithFacebook>;
    signInWithGoogle: ReturnType<typeof signInWithGoogle>;
}

export interface TokenOptions {
    lifetimeMinutes?: number;
    refreshLifetimeMinutes?: number;
    jwtSecret?: string;
    jwtPublicKey?: string;
    jwtPrivateKey?: string;
}

export interface User {
    displayName?: string;
    email?: string;
    emailVerified?: boolean;
    isAnonymous?: boolean;
    metadata?: UserMetadata;
    phoneNumber?: string;
    photoUrl?: string;
    providerData: UserInfo;
    uid: string;
}

export interface UserInfo {
    email?: string;
    displayName?: string;
    phoneNumber?: string;
    providerId?: string;
    uid?: string;
}

export interface UserMetadata {
    creationTime: Date;
    lastSignInTime: Date;
}

export interface AuthistOptions {
    emailPassword?: EmailPasswordProviderOptions;
    facebook?: FacebookProviderOptions;
    google?: GoogleSignInOptions;
    onAuthenticationFailure?: (error: Error, req: any) => void;
    onExpressAuthenticationFailure?: (error: Error, req: Request, res: Response, next: NextFunction) => void;
    sendRegistrationEmail?: (user: User, req: any) => Promise<void>;
    token?: TokenOptions;
    getUserById: (uid: string) => Promise<User | undefined>;
}

export interface UserCredentials {
    user: User;
    credentials: Credentials;
}

export interface Credentials {
    accessToken: string;
    refreshToken: string;
    expiresIn: number;
    refreshExpiresIn: number;
}
