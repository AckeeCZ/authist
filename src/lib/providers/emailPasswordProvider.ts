import { compareSync, hashSync } from 'bcrypt';
import { AuthistOptions, User, UserCredentials, UserInfo } from '../authist';
import { createCredentials } from '../credentialsService';
import { ERROR_CODE, NotAuthenticated } from '../error';
import { handleError } from '../providerUtils';

export enum HashingAlgorithm {
    Bcrypt = 'bcrypt',
    Plaintext = 'plaintext',
}

interface SignInParams {
    email: string;
    password: string;
    request: any;
}

interface ValidatePasswordOptions {
    password: string;
    hashedPassword: string;
    request: any;
}

export interface EmailPasswordProviderOptions {
    passwordHashingAlgorithm?: HashingAlgorithm;
    getUserByEmail: (email: string, req: any) => Promise<(User & { password: string }) | undefined>;
    saveNonExistingUser?: (data: UserInfo, hashedPassword: string, req: any) => Promise<User>;
    sendResetPasswordEmail?: (user: User, req: any) => Promise<void>;
    validateUser?: (data: UserInfo, password: string, req: any) => Promise<void>;
    validateSignIn?: (email: string, password: string, req: any) => Promise<void>;
}

export const signInWithEmailAndPassword = (options: AuthistOptions) => async (
    email: string,
    password: string,
    req?: any
): Promise<UserCredentials> => {
    try {
        const emailPasswordOptions = options.emailPassword!;
        let user = await emailPasswordOptions.getUserByEmail(email, req);
        if (!user) {
            if (!emailPasswordOptions.saveNonExistingUser) {
                throw new NotAuthenticated(ERROR_CODE.UserNotFound);
            }
            await validateUser({ email, password, request: req }, options);
            user = await saveNonExistingUser({ email, password, request: req }, options);
            await sendRegistrationEmail(user, req, options);
        }
        await validateSignIn({ email, password, request: req }, options);
        await validatePassword({ password, request: req, hashedPassword: user.password }, options);
        const { password: _psw, ...rest } = user;
        const credentials = createCredentials(user, options);
        return { credentials, user: rest };
    } catch (error) {
        return handleError(error, req, options);
    }
};

export const getHashingAlgorithm = (options?: EmailPasswordProviderOptions) => {
    if (options?.passwordHashingAlgorithm) {
        return options.passwordHashingAlgorithm;
    }
    return HashingAlgorithm.Bcrypt;
};

export const validatePassword = async (params: ValidatePasswordOptions, options: AuthistOptions) => {
    try {
        const algorithm = getHashingAlgorithm(options.emailPassword);
        if (algorithm === HashingAlgorithm.Plaintext && params.password !== params.hashedPassword) {
            throw new NotAuthenticated(ERROR_CODE.PasswordMismatch);
        }
        if (algorithm === HashingAlgorithm.Bcrypt && !compareSync(params.password, params.hashedPassword)) {
            throw new NotAuthenticated(ERROR_CODE.PasswordMismatch);
        }
    } catch (error) {
        return handleError(error, params.request, options);
    }
};

export const validateSignIn = async (params: SignInParams, options: AuthistOptions) => {
    try {
        await (options.emailPassword?.validateSignIn
            ? options.emailPassword.validateSignIn(params.email, params.password, params.request)
            : Promise.resolve());
    } catch (error) {
        return handleError(error, params.request, options);
    }
};

export const sendRegistrationEmail = async (user: User, request: any, options: AuthistOptions) => {
    try {
        await (options.sendRegistrationEmail ? options.sendRegistrationEmail(user, request) : Promise.resolve());
    } catch (error) {
        return handleError(error, request, options);
    }
};

export const saveNonExistingUser = async (
    params: SignInParams,
    options: AuthistOptions
): Promise<User & { password: string }> => {
    try {
        const hashedPassword = hashPassword(params.password, options.emailPassword!);
        const userInfo = getUserInfo(params);
        await (options.emailPassword?.saveNonExistingUser
            ? options.emailPassword.saveNonExistingUser(userInfo, hashedPassword, params.request)
            : Promise.resolve({} as User));
        const user = await options.emailPassword?.getUserByEmail(params.email, params.request);
        return user!;
    } catch (error) {
        return handleError(error, params.request, options) as any;
    }
};

const hashPassword = (password: string, options: EmailPasswordProviderOptions) => {
    const algorithm = getHashingAlgorithm(options);
    if (algorithm === HashingAlgorithm.Bcrypt) {
        return hashSync(password, 12);
    }
    return password;
};

export const validateUser = async (params: SignInParams, options: AuthistOptions) => {
    try {
        const userInfo = getUserInfo(params);
        await (options.emailPassword?.validateUser
            ? options.emailPassword.validateUser(userInfo, params.password, params.request)
            : Promise.resolve());
    } catch (error) {
        return handleError(error, params.request, options);
    }
};

const getUserInfo = (params: SignInParams): UserInfo => ({
    email: params.email,
});
