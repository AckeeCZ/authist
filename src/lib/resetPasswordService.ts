import { sign, verify } from 'jsonwebtoken';
import { AuthistOptions, User } from './authist';
import { getSecret } from './credentialsService';
import { ERROR_CODE, NotAuthenticated } from './error';
import { hashPassword } from './providers/emailPasswordProvider';
import { handleError } from './providerUtils';

const RESET_TOKEN_TYPE = 'resetPasswordToken';

export const resetPassword = (options: AuthistOptions) => async (email: string, req?: any) => {
    try {
        const opts = options.emailPassword!;
        const user = await opts.getUserByEmail(email, req);
        if (!user) {
            throw new NotAuthenticated(ERROR_CODE.UserNotFound);
        }
        const token = await getResetToken(options, user);
        await (opts?.sendResetPasswordEmail ? opts.sendResetPasswordEmail(token, user, req) : Promise.resolve());
        await (opts.saveResetPasswordToken ? opts.saveResetPasswordToken(token, user, req) : Promise.resolve());
        return { token };
    } catch (error) {
        return handleError(error, req, options);
    }
};

export const changePassword = (options: AuthistOptions) => async (token: string, password: string, req?: any) => {
    try {
        const opts = options.emailPassword!;
        if (!opts.updatePassword) {
            throw new NotAuthenticated(ERROR_CODE.UpdatePasswordNotImplemented);
        }
        const user = await validateResetToken(token, options, req);
        if (!user) {
            throw new NotAuthenticated(ERROR_CODE.UserNotFound);
        }
        await (opts.validatePassword ? opts.validatePassword(password, req) : Promise.resolve());
        const hashedPassword = hashPassword(password, opts);
        await (opts.updatePassword ? opts.updatePassword(hashedPassword, user, req) : Promise.resolve());
    } catch (error) {
        return handleError(error, req, options);
    }
};

const validateResetToken = async (token: string, options: AuthistOptions, req?: any) => {
    try {
        if (options.emailPassword?.validateResetPasswordToken) {
            return await options.emailPassword.validateResetPasswordToken(token, req);
        }
        const secret = getSecret(options, true);
        const decodedToken = verify(token, secret) as { email: string; type: string };
        if (decodedToken.type !== RESET_TOKEN_TYPE) {
            throw new NotAuthenticated(ERROR_CODE.InvalidResetPasswordTokenType);
        }
        if (!decodedToken.email) {
            throw new NotAuthenticated(ERROR_CODE.InvalidResetPasswordToken);
        }
        return options.emailPassword!.getUserByEmail(decodedToken.email, req);
    } catch (error) {
        return handleError(error, req, options);
    }
};

export const getResetToken = async (options: AuthistOptions, user: User) => {
    if (!options!.emailPassword?.getResetPasswordToken) {
        return sign({ type: RESET_TOKEN_TYPE, email: user.email }, getSecret(options), {
            expiresIn: (options.token?.lifetimeMinutes || 60) * 60,
        });
    }
    return options!.emailPassword!.getResetPasswordToken!(user);
};
