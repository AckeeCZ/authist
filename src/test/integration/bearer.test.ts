import express from 'express';
import got from 'got';
import { Server } from 'http';
import { AddressInfo } from 'net';
import { AuthistOptions, createAuthenticator, emailPasswordProvider, ErrorCode, User } from '../../lib';
import { HashingAlgorithm } from '../../lib/providers/emailPasswordProvider';
import { getResetToken } from '../../lib/resetPasswordService';
import { getUserByEmail, getUserById, initDb, updateUser } from '../databaseUtils';

let userModel: any;
let app: Server;
const getAuthenticator = (options: AuthistOptions) => createAuthenticator(options);

describe('Bearer authentication middleware', () => {
    const startServer = (options: AuthistOptions) => {
        const authenticator = getAuthenticator(options);
        const server = express();
        server.get('/', authenticator.expressBearer, (req: any, res) => res.json(req.user));
        server.get('/refresh-token', authenticator.expressRefreshToken);
        app = server.listen(0);
        return `http://0.0.0.0:${(app.address()! as AddressInfo).port}`;
    };
    const getAuthistOptions = (): AuthistOptions => ({
        getUserById: getUserById(userModel),
        emailPassword: {
            getUserByEmail: getUserByEmail(userModel),
            saveNonExistingUser: (data, hashedPassword, _req) =>
                userModel.create({ email: data.email, password: hashedPassword }),
        },
    });
    const email = 'tester@test.com';
    const password = '***';
    beforeAll(async () => {
        userModel = await initDb();
        const user = await emailPasswordProvider.saveNonExistingUser(
            { email, password, request: {} },
            getAuthistOptions()
        );
        expect(user).not.toBeNull();
    });
    afterEach(() => app.close());
    afterAll(() => userModel.dropAllSchemas());
    test('Unauthorized return an error', async () => {
        const serverUrl = startServer(getAuthistOptions());
        try {
            await got(serverUrl, { responseType: 'json' });
        } catch (error) {
            expect(error.response.statusCode).toBe(401);
            return expect(error.response.body.message).toBe(ErrorCode.AuthenticationRequired);
        }
        throw new Error('Expected to throw');
    });
    test('Unsupported type return an error', async () => {
        const serverUrl = startServer(getAuthistOptions());
        try {
            await got(serverUrl, {
                responseType: 'json',
                headers: { authorization: 'basic 123' },
            });
        } catch (error) {
            expect(error.response.statusCode).toBe(401);
            return expect(error.response.body.message).toBe(ErrorCode.UnsupportedAuthorization);
        }
        throw new Error('Expected to throw');
    });
    test('Invalid token return an error', async () => {
        const serverUrl = startServer(getAuthistOptions());
        try {
            await got(serverUrl, {
                responseType: 'json',
                headers: { authorization: 'bearer invalid_token' },
            });
        } catch (error) {
            expect(error.response.statusCode).toBe(401);
            return expect(error.response.body.message).toMatchSnapshot();
        }
        throw new Error('Expected to throw');
    });
    test('Invalid token return custom error', async () => {
        const serverUrl = startServer({
            ...getAuthistOptions(),
            onAuthenticationFailure: () => {
                throw new Error(customError);
            },
        });
        const customError = 'My custom error';
        try {
            await got(serverUrl, {
                responseType: 'json',
                headers: { authorization: 'bearer invalid_token' },
            });
        } catch (error) {
            expect(error.response.statusCode).toBe(401);
            return expect(error.response.body.message).toBe(customError);
        }
        throw new Error('Expected to throw');
    });
    test('Can use onExpressAuthenticationFailure', async () => {
        const serverUrl = startServer({
            ...getAuthistOptions(),
            onAuthenticationFailure: () => {
                throw new Error(customError);
            },
            onExpressAuthenticationFailure: (error, _, res) => {
                res.send(JSON.stringify(error, Object.getOwnPropertyNames(error)));
            },
        });
        const customError = 'My custom error';
        const { body } = await got<{ message: string }>(serverUrl, {
            responseType: 'json',
            headers: { authorization: 'bearer invalid_token' },
        });
        expect(Object.keys(body).sort()).toStrictEqual(['message', 'stack'].sort());
        expect(body.message).toBe(customError);
    });
    test('Can authorize user', async () => {
        const options = getAuthistOptions();
        const serverUrl = startServer(options);
        const { credentials } = await emailPasswordProvider.signInWithEmailAndPassword(options)(email, password);
        const { body } = await got<User & { password?: string }>(serverUrl, {
            responseType: 'json',
            headers: { authorization: `Bearer ${credentials.accessToken}` },
        });
        expect(body.password).toBeUndefined();
        expect(body.email).toBe(email);
        expect(body).toMatchSnapshot();
    });
    test('Can refresh token', async () => {
        const options = getAuthistOptions();
        const serverUrl = startServer(options);
        const { credentials } = await emailPasswordProvider.signInWithEmailAndPassword(options)(email, password);
        const { body } = await got<User & { password?: string }>(`${serverUrl}/refresh-token`, {
            responseType: 'json',
            searchParams: { refreshToken: credentials.refreshToken },
        });
        expect(Object.keys(body).sort()).toStrictEqual(['user', 'credentials'].sort());
    });
    describe('Reset password', () => {
        const startResetPasswordServer = (options: AuthistOptions) => {
            const authenticator = getAuthenticator(options);
            const server = express();
            server.get('/', async (req: any, res) => {
                const data = { ...req.body, ...req.query };
                const credentials = await authenticator.signInWithEmailAndPassword(data.email, data.password, req);
                res.json(credentials);
            });
            server.get('/recover-password', authenticator.expressRecoverPassword);
            server.get('/reset-password', authenticator.expressResetPassword);
            app = server.listen(0);
            return `http://0.0.0.0:${(app.address()! as AddressInfo).port}`;
        };
        afterEach(() => app.close());
        test('User can recover password', async () => {
            const serverUrl = startResetPasswordServer(getRecoverPasswordOptions());
            const { statusCode } = await got<{ token: string }>(`${serverUrl}/recover-password`, {
                searchParams: { email },
                responseType: 'json',
            });
            expect(statusCode).toBe(200);
        });
        test('Non-existing user cannot recover password', async () => {
            try {
                const serverUrl = startResetPasswordServer(getRecoverPasswordOptions());
                await got(`${serverUrl}/recover-password`, {
                    searchParams: { token: 'bad-token' },
                });
                throw new Error('Expected to throw');
            } catch (error) {
                expect(error.response.statusCode).toBe(401);
            }
        });
        test('User can reset password', async () => {
            const options = getResetPasswordOptions();
            const url = startResetPasswordServer(options);
            const psw = 'newPassword';
            const token = await getResetToken(options, { email } as any);
            await got(`${url}/reset-password`, {
                searchParams: { token, password: psw },
            });
            const { body } = await got<{ user: User }>(url, {
                searchParams: { email, password: psw },
                responseType: 'json',
            });
            expect(Object.keys(body).sort()).toStrictEqual(['user', 'credentials'].sort());
            expect(body.user.email).toBe(email);
        });
        test('Non-existing user cannot reset password', async () => {
            try {
                const url = startResetPasswordServer(getResetPasswordOptions());
                await got(`${url}/reset-password`, {
                    searchParams: { password, token: 'bad-token' },
                });
                throw new Error('Expected to throw');
            } catch (error) {
                expect(error.response.statusCode).toBe(400);
            }
        });
    });
});

const getRecoverPasswordOptions = () => ({
    getUserById: getUserById(userModel),
    emailPassword: {
        getUserByEmail: getUserByEmail(userModel),
    },
});

const getResetPasswordOptions = () => ({
    getUserById: getUserById(userModel),
    emailPassword: {
        getUserByEmail: getUserByEmail(userModel),
        passwordHashingAlgorithm: HashingAlgorithm.Plaintext,
        updatePassword: (password: string, user: User) => updateUser(userModel)({ password }, user),
    },
});
