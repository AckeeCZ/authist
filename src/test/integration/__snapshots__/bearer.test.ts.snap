// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bearer authentication middleware Can authorize user 1`] = `
Object {
  "email": "tester@test.com",
  "providerData": Object {},
  "uid": "1",
}
`;

exports[`Bearer authentication middleware Invalid token return an error 1`] = `"jwt malformed"`;
